### PR TITLE
HEEDLS-393 Add profile picture and information from database

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
Added the profile image to the My Details expander and converted the data shown on the my account page to all be retrieved from the database using only the Admin Id, Candidate Id and Centre Id from the Claims. The same default profile image is shown as the old site when there is no user profile image.

I've also refactored the UserService into a UserService and a UserDataService. The UserDataService contains all the methods that actually call the database, and the UserService contains logic which uses those. This allows more of the logic in the UserService to be unit tested easily.

Checked that the layout seems fine in Chrome and Chrome mobile devices emulation. Also ran all unit tests (and added some new ones)

image

### CODE REVIEW

File: DigitalLearningSolutions.Web.Tests/TestHelpers/MyAccountHelper.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web.Tests/TestHelpers/MyAccountHelper.cs:7]: Are these GetDefault...User not already present in the UserTestHelper in the data.tests project, or is there a benefit to having them in both test projects?

File: DigitalLearningSolutions.Web.Tests/ViewModels/MyAccount/MyAccountViewModelTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web.Tests/ViewModels/MyAccount/MyAccountViewModelTests.cs:28]: There are a lot of assertions here to test each individual property. Do we think an Assert.Multiple() might be of benefit? https://docs.nunit.org/articles/nunit/writing-tests/assertions/multiple-asserts.html
The docs suggest it would allow multiple errors to be displayed at once, rather than terminating on the first, but I don't know if that's how the present code already works - or as they are testing what should be the same model if one property is wrong we can assume the rest will be wrong as well.

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:35]: Given that we need to call this method twice, with different usernames, could we split it into two different tests?

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:55]: Could we just do var (returnedAdminUser, returnedDelegateUser) and call userService once?

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:35]: Unsure, I didn't write the initial unit test, The only change I made to it was because I added the userDataService dependency for the methods it calls. I can look to change it

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:55]: Yeah probably, I just copied the pattern from the other existing tests I refactored slightly

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:35]: I think I initially wrote this test and it was just modified as part of the changes here. While it was originally written to test the scenario of a function returning both in one call, because of the way the test users are set up, they cannot both be returned at once. So I suppose splitting it into two might be best.

File: DigitalLearningSolutions.Web/Controllers/MyAccountController.cs
Comment (stellake) [inline DigitalLearningSolutions.Web/Controllers/MyAccountController.cs:34]: I wonder if we even need the centre name anymore as it looks like we're joining with the centres table in the db queries when getting the user data..

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Controllers/MyAccountController.cs:34]: Yeah we shouldn't need to, particularly as I need to get more information off the centres for the third subtask

File: DigitalLearningSolutions.Data/Models/User/User.cs
Comment (stellake) [inline DigitalLearningSolutions.Data/Models/User/User.cs:21]: Isn't this declared as nullable on the view model?

File: DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:69]: I'm a little bit nervous doing this by id because it will most likely break for some people locally (as the add test users script can be ran on the db backup at different times). At the same time, I don't want any password hashes being hardcoded in the codebase. 
Do you think we could potentially fetch the users based on an id on the db backup (e.g. 3, which will be the same for everyone) and then check that the fields except for the password hash are as expected? If doing it is a bit messy, we can just check that the fields we care about are as expected (the ones you'll be using on the My account page.

Comment (stellake) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:69]: Or to be honest, we could just create a script that replaces one of the user password hashes that is in the db backup with "password" and then compare against that (which I think is even better than creating/using the new test users) 👀  ?

Comment (stellake) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:69]: Any ideas/opinions? :)

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:69]: Yeah didn't think about whether it would break on other people's environments. It would definitely be messy (I guess that is to be expected as these aren't really unit tests) but I could write tests that query against known users from the db backup. Would end up hardcoding in a load of expected user values with the details I need to check instead.

Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:69]: I don't believe any password hashes are currently stored in the codebase. The UserTestHelper defaults just have "password" as the password by default, and I don't think that is overwritten anywhere. The test user that that is based on also just has password as "password". If one of the records in the db backup (prior to migrations/scripts), can be safely modified to be the new test user, I don't see any problem with, and is almost certainly preferable to the new test user with the potentially varying Id.

Comment (stellake) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:69]: Cool - so maybe we can replace the test user script with a different script that just changes the pw of a single candidate and single admin record that are both present in the db backup (and maybe modify some other fields we might need)? Then, we can make default users have these fields and use the same users for all of these data tests.

File: DigitalLearningSolutions.Web.Tests/ViewModels/MyAccount/MyAccountViewModelTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web.Tests/ViewModels/MyAccount/MyAccountViewModelTests.cs:28]: Looked into this, apparently FluentAssertions don't work with the NUnit Assert.Multiple, but it has a similar AssertionScope that I'll use instead.

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:35]: Since this test is no longer testing against the database, it doesn't actually matter what string we are inputting since the fake UserDataService is accepting A<string>._. Combined it into a single call in a single test. There are separate tests in the UserDataServiceTests that test against the actual data

File: DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs]: I've updated the PopulateTestUsers script to instead update the first record (other than the admin record ID 1) of each table with the properties we were setting on the created users. These are AdminID 7 and CandidateID 1.
I've also added an update statement for the CentreID 2 so we can set a few of the properties on there as needed.
We should now be able to test against these users/centre for all of the tests we need to, amending the update statements with further data as necessary (e.g. I'll be adding values to the AnswerX fields for 392)

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs]: This may mean we need to remove the created test users from any databases where they've been added

File: DigitalLearningSolutions.Data.Tests/Helpers/UserTestHelper.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Helpers/UserTestHelper.cs:83]: Could this be a variable?

File: DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:37]: Super minor, but should this be singular "User"

File: DigitalLearningSolutions.Data.Tests/Helpers/UserTestHelper.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/Helpers/UserTestHelper.cs:83]: Ah whoops I intended to delete that, it's not used any more

File: DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs:60]: ditto about the singular of "user"

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:30]: Should these return "expected admin user" etc (across all the tests)?

File: DigitalLearningSolutions.Web/Views/MyAccount/Index.cshtml
Comment (stellake) [inline DigitalLearningSolutions.Web/Views/MyAccount/Index.cshtml:107]: Just a thought and I'm not 100% sure, but could "Default profile picture" indicate that this will be the user's profile picture if they don't upload anything? Would it be worth naming it "No profile picture"?

**Review by stellake — APPROVED** [review]
A few really small comments - happy for you to merge once these are addressed. Don't forget to post in the hee-internal group for people to run the new migration after you've pressed the merge button.

Well done refactoring the userService into service & data service! 👍

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:30]: I think that they can be, as these should be testing what the UserService functions return rather than what is passed in to the UserService function, so it should be safe to reuse the expected variable. So I think that might be a better way forward.

File: DigitalLearningSolutions.Web/Views/MyAccount/Index.cshtml
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/MyAccount/Index.cshtml:107]: Unsure, I guess it depends how the rest of the site and old behaves when a user doesn't have a set profile picture, do they always display with the default? Do we not display a profile picture? I went with default as it implied to me it was a non-user set profile picture.

**Pull request merged**
